### PR TITLE
8293562: KeepAliveCache Blocks Threads while Closing Connections

### PR DESCRIPTION
**Pull request by dhanalla**
Issue: [JDK-8293562](https://bugs.openjdk.org/browse/JDK-8293562): KeepAliveCache Blocks Threads while Closing Connections
Backport of PR https://github.com/openjdk/jdk11u-dev/pull/1890 fromJDK11 to JDK8.

The JDK11 patch doesn't apply cleanly 3/11 hunks need to be ported manually in KeepAliveCache.java.

Test performed:

1. JTreg jdk_tier1
2. JTreg jdk_tier2
3. New test case added to this PR passed with the fix and failed without it.
4. The customer failure scenario has been validated.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] [JDK-8293562](https://bugs.openjdk.org/browse/JDK-8293562) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293562](https://bugs.openjdk.org/browse/JDK-8293562): KeepAliveCache Blocks Threads while Closing Connections (**Bug** - P4 - Approved)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/409/head:pull/409` \
`$ git checkout pull/409`

Update a local copy of the PR: \
`$ git checkout pull/409` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/409/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 409`

View PR using the GUI difftool: \
`$ git pr show -t 409`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/409.diff">https://git.openjdk.org/jdk8u-dev/pull/409.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/409#issuecomment-1861388673)

### CODE REVIEW

**Comment by dhanalla** [comment]
> @dhanalla This pull request has been inactive for more than 4 weeks and will be automatically closed if another 4 weeks passes without any activity. To avoid this, simply add a new comment to the pull request. Feel free to ask for assistance if you need help with progressing this pull request towards integration!

adding a comment to keep the PR active.

File: jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java
Comment (jerboaa) [inline jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java:270]: Indenting is off.

File: jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java
Comment (jerboaa) [inline jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java:231]: There are some whitespace issues here. Please clean this up.

File: jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java
Comment (jerboaa) [inline jdk/src/share/classes/sun/net/www/http/KeepAliveCache.java:147]: Whitespace and aligned-ness of braces are off here.

**Review by jerboaa — COMMENTED** [review]
This PR seems OK. It would have been easier to review if the whitespace wasn't all different. Consider adding https://github.com/jerboaa/jdk8u-dev/commit/054efe518681e6334b700c8a236fa4c508db640d to make this more aligned with JDK 11u. Finally, I wonder if we should bring in https://bugs.openjdk.org/browse/JDK-8278067 first so as to make the patch cleaner (as compared to JDK 11). That one is in 8u361 oracle as well, so would probably good to have and seems low risk. Meanwhile, please apply for approval for #407 so this can get in.

**Comment by dhanalla** [comment]
> This PR seems OK. It would have been easier to review if the whitespace wasn't all different. Consider adding [jerboaa@054efe5](https://github.com/jerboaa/jdk8u-dev/commit/054efe518681e6334b700c8a236fa4c508db640d) to make this more aligned with JDK 11u. Finally, I wonder if we should bring in https://bugs.openjdk.org/browse/JDK-8278067 first so as to make the patch cleaner (as compared to JDK 11). That one is in 8u361 oracle as well, so would probably good to have and seems low risk. Meanwhile, please apply for approval for #407 so this can get in.

Thanks @jerboaa for reviewing this PR. 
I have submitted a request for approval for PR https://github.com/openjdk/jdk8u-dev/pull/407 and have addressed the indentation and spacing issues.

**Comment by dhanalla** [comment]
> > This PR seems OK. It would have been easier to review if the whitespace wasn't all different. Consider adding [jerboaa@054efe5](https://github.com/jerboaa/jdk8u-dev/commit/054efe518681e6334b700c8a236fa4c508db640d) to make this more aligned with JDK 11u. Finally, I wonder if we should bring in https://bugs.openjdk.org/browse/JDK-8278067 first so as to make the patch cleaner (as compared to JDK 11). That one is in 8u361 oracle as well, so would probably good to have and seems low risk. Meanwhile, please apply for approval for #407 so this can get in.
> 
> Thanks @jerboaa for reviewing this PR. I have submitted a request for approval for PR #407 and have addressed the indentation and spacing issues.

I am happy to backport the bug https://bugs.openjdk.org/browse/JDK-8278067, but my concern is that the backport of the other files in the PR (https://github.com/openjdk/jdk11u-dev/pull/1193) may not be clean and could result in bringing in other changes.

**Comment by theRealAph** [comment]
I'm surprised you think this is suitable for 8u. It's not low risk, and it's more of a performance fix than a bug, and it's P4.
JDK 8u is done. It's a long-term maintenance release.

**Comment by dhanalla** [comment]
> I'm surprised you think this is suitable for 8u. It's not low risk, and it's more of a performance fix than a bug, and it's P4. JDK 8u is done. It's a long-term maintenance release.

Thanks for reviewing this PR @theRealAph ,

I understand that this is not a Low-risk. We've observed that two of our valued customers' applications, which heavily rely on HttpClient and handle thousands of requests per second, encounter significant challenges with TLS 1.3. Specifically, these applications become almost unresponsive due to a huge number of blocked threads (The same issue is explained here https://techcommunity.microsoft.com/t5/azure-storage-blog/prepare-for-upcoming-tls-1-3-support-for-azure-storage/ba-p/4034014) Unfortunately, due to various constraints, these customers are unable to upgrade to the later JDK version or revert to TLS 1.2. Both of these customers have explicitly conveyed that implementing this fix would greatly alleviate their current roadblocks.
 
The attached test simulates a scenario in which the second HTTP client read request fails with a timeout as it is blocked by the first socket close connection. This blocking occurs due to synchronization locks in the KeepAliveCache's put and Keep-Alive-Timer thread methods, with the Keep-Alive-Timer thread handling socket closure while holding a lock. This means that while it's closing connections, no other connections can be established, retrieved from the cache, or added to the cache. And the test passes with the fix.

**Comment by jerboaa** [comment]
If we are going to backport this, then I think we should use the following approach in order to keep the risk manageable (and to keep in sync with what had been done in JDK 11):

1. Get `JDK-8278067: Make HttpURLConnection default keep alive timeout configurable` into JDK 8u. That's #437 
2. Get `JDK-8291226: Create Test Cases to cover scenarios for JDK-8278067` backported as well as it includes correctness tests for `JDK-8293562` as well.
3. Get `JDK-8291638: Keep-Alive timeout of 0 should close connection immediately` backported.

**Comment by dhanalla** [comment]
> If we are going to backport this, then I think we should use the following approach in order to keep the risk manageable (and to keep in sync with what had been done in JDK 11):
> 
> 1. Get `JDK-8278067: Make HttpURLConnection default keep alive timeout configurable` into JDK 8u. That's [8278067: Make HttpURLConnection default keep alive timeout configurable #437](https://github.com/openjdk/jdk8u-dev/pull/437)
> 2. Get `JDK-8291226: Create Test Cases to cover scenarios for JDK-8278067` backported as well as it includes correctness tests for `JDK-8293562` as well.
> 3. Get `JDK-8291638: Keep-Alive timeout of 0 should close connection immediately` backported.

Thanks @jerboaa , 
created backport PRs for both 2 & 3.
JDK-8291226: https://github.com/openjdk/jdk8u-dev/pull/453
JDK-8291638: https://github.com/openjdk/jdk8u-dev/pull/454

**Comment by ltning** [comment]
I'd like to pitch in that this feels (to us) like a serious bug that warrants fixing in 8. I don't know what the criteria for calling something a "performance issue" is, but my gut feeling tells me it falls far short of being classified as "just" that. Our application ground to a complete halt (not just a slowdown) when the client switched on TLS 1.3, and it took us the better part of a day to figure out the root cause and restore service.

My few cents, and a huge thanks to @dhanalla and anyone else working on backporting this fix.

**Review by jerboaa — COMMENTED** [review]
This seems OK now modulo some whitespace changes. The best course of action would be to compare the `KeepAliveCache.java` file to the version from `jdk11u` and fix the whitespace changes in the  current version. The only difference then should be in the `AccessController.doPrivileged(` block between JDK 8 and 11 and nothing else.

http tests pass for me:
```
Passed: sun/net/www/http/ChunkedInputStream/ChunkedEncodingTest.java
Passed: sun/net/www/http/ChunkedInputStream/ChunkedEncodingWithProgressMonitorTest.java
Passed: sun/net/www/http/ChunkedInputStream/TestAvailable.java
Passed: sun/net/www/http/ChunkedOutputStream/checkError.java
Passed: sun/net/www/http/ChunkedOutputStream/Test.java
Passed: sun/net/www/http/HttpClient/B6726695.java
Passed: sun/net/www/http/HttpClient/B7025238.java
Passed: sun/net/www/http/HttpClient/B8025710.java
Passed: sun/net/www/http/HttpClient/B8209178.java
Passed: sun/net/www/http/HttpClient/CookieHttpClientTest.java
Passed: sun/net/www/http/HttpClient/GetProxyPort.java
Passed: sun/net/www/http/HttpClient/ImplicitFileName.java
Passed: sun/net/www/http/HttpClient/IsAvailable.java
Passed: sun/net/www/http/HttpClient/IsKeepingAlive.java
Passed: sun/net/www/http/HttpClient/KeepAliveTest.java
Passed: sun/net/www/http/HttpClient/MultiThreadTest.java
Passed: sun/net/www/http/HttpClient/OpenServer.java
Passed: sun/net/www/http/HttpClient/ProxyFromCache.java
Passed: sun/net/www/http/HttpClient/ProxyTest.java
Passed: sun/net/www/http/HttpClient/RequestURI.java
Passed: sun/net/www/http/HttpClient/RetryPost.sh
Passed: sun/net/www/http/HttpClient/StreamingRetry.java
Passed: sun/net/www/http/HttpURLConnection/PostOnDelete.java
Passed: sun/net/www/http/KeepAliveCache/B5045306.java
Passed: sun/net/www/http/KeepAliveCache/B8291637.java
Passed: sun/net/www/http/KeepAliveCache/B8293562.java
Passed: sun/net/www/http/KeepAliveCache/KeepAliveProperty.java
Passed: sun/net/www/http/KeepAliveCache/KeepAliveTimerThread.java
Passed: sun/net/www/http/KeepAliveStream/InfiniteLoop.java
Passed: sun/net/www/http/KeepAliveStream/KeepAliveStreamClose.java
Passed: sun/net/www/http/KeepAliveStream/KeepAliveStreamCloseWithWrongContentLength.java
Passed: sun/net/www/http/RequestMethodCheck/RequestMethodEquality.java
Test results: passed: 32
```

**Comment by dhanalla** [comment]
> This seems OK now modulo some whitespace changes. The best course of action would be to compare the `KeepAliveCache.java` file to the version from `jdk11u` and fix the whitespace changes in the current version. The only difference then should be in the `AccessController.doPrivileged(` block between JDK 8 and 11 and nothing else.
> 
> http tests pass for me:
> 
> ```
> Passed: sun/net/www/http/ChunkedInputStream/ChunkedEncodingTest.java
> Passed: sun/net/www/http/ChunkedInputStream/ChunkedEncodingWithProgressMonitorTest.java
> Passed: sun/net/www/http/ChunkedInputStream/TestAvailable.java
> Passed: sun/net/www/http/ChunkedOutputStream/checkError.java
> Passed: sun/net/www/http/ChunkedOutputStream/Test.java
> Passed: sun/net/www/http/HttpClient/B6726695.java
> Passed: sun/net/www/http/HttpClient/B7025238.java
> Passed: sun/net/www/http/HttpClient/B8025710.java
> Passed: sun/net/www/http/HttpClient/B8209178.java
> Passed: sun/net/www/http/HttpClient/CookieHttpClientTest.java
> Passed: sun/net/www/http/HttpClient/GetProxyPort.java
> Passed: sun/net/www/http/HttpClient/ImplicitFileName.java
> Passed: sun/net/www/http/HttpClient/IsAvailable.java
> Passed: sun/net/www/http/HttpClient/IsKeepingAlive.java
> Passed: sun/net/www/http/HttpClient/KeepAliveTest.java
> Passed: sun/net/www/http/HttpClient/MultiThreadTest.java
> Passed: sun/net/www/http/HttpClient/OpenServer.java
> Passed: sun/net/www/http/HttpClient/ProxyFromCache.java
> Passed: sun/net/www/http/HttpClient/ProxyTest.java
> Passed: sun/net/www/http/HttpClient/RequestURI.java
> Passed: sun/net/www/http/HttpClient/RetryPost.sh
> Passed: sun/net/www/http/HttpClient/StreamingRetry.java
> Passed: sun/net/www/http/HttpURLConnection/PostOnDelete.java
> Passed: sun/net/www/http/KeepAliveCache/B5045306.java
> Passed: sun/net/www/http/KeepAliveCache/B8291637.java
> Passed: sun/net/www/http/KeepAliveCache/B8293562.java
> Passed: sun/net/www/http/KeepAliveCache/KeepAliveProperty.java
> Passed: sun/net/www/http/KeepAliveCache/KeepAliveTimerThread.java
> Passed: sun/net/www/http/KeepAliveStream/InfiniteLoop.java
> Passed: sun/net/www/http/KeepAliveStream/KeepAliveStreamClose.java
> Passed: sun/net/www/http/KeepAliveStream/KeepAliveStreamCloseWithWrongContentLength.java
> Passed: sun/net/www/http/RequestMethodCheck/RequestMethodEquality.java
> Test results: passed: 32
> ```

Done

**Review by jerboaa — APPROVED** [review]
This looks OK to me, but I'd like for somebody else to give the go-ahead as well.

**Review by phohensee — APPROVED** [review]
Lgtm. The only real difference from the 11u backport looks to be the keepAliveTimer thread creation code.

**Comment by dhanalla** [comment]
Thanks @jerboaa & @phohensee  for reviewing this PR.
I added the JDK8u fix request tag to the bug as the next step. Please let me know if there is anything else that needs to be addressed.

**Comment by gnu-andrew** [comment]
> This seems OK now modulo some whitespace changes. The best course of action would be to compare the `KeepAliveCache.java` file to the version from `jdk11u` and fix the whitespace changes in the current version. The only difference then should be in the `AccessController.doPrivileged(` block between JDK 8 and 11 and nothing else.
> 

This now looks correct after applying the patch, though the 8u & 11u patches look different to achieve the same end result. There's an additional whitespace change which I'm not sure if it was missed in an earlier backport or 8u only to begin with, but the file looks good after this patch, which is the main thing.

**Review by gnu-andrew — APPROVED** [review]
Backport looks good.

**Comment by dhanalla** [comment]
> @gnu-andrew 8293562: The approval request has been approved.

Thanks @gnu-andrew for reviewing this PR.

**Pull request closed without merging**
